core/services/keeper/upkeep_executer.go

```diff
@@ -0,0 +1,189 @@
+package keeper
+
+import (
+	"context"
+	"errors"
+	"math/big"
+
+	"github.com/ethereum/go-ethereum"
+	"github.com/smartcontractkit/chainlink/core/logger"
+	"github.com/smartcontractkit/chainlink/core/services/eth"
+	"github.com/smartcontractkit/chainlink/core/store"
+	"github.com/smartcontractkit/chainlink/core/store/models"
+	"github.com/smartcontractkit/chainlink/core/utils"
+	"go.uber.org/atomic"
+	"gorm.io/gorm"
+)
+
+const (
+	checkUpkeep        = "checkUpkeep"
+	performUpkeep      = "performUpkeep"
+	executionQueueSize = 10
+)
+
+func NewUpkeepExecuter(
+	db *gorm.DB,
+	ethClient eth.Client,
+) UpkeepExecuter {
+	return UpkeepExecuter{
+		blockHeight:    atomic.NewInt64(0),
```

Comment:
I'd imagine we probably don't care about making this thread-safe. OnNewLongestChain should never be called concurrently, that would violate a bunch of assumptions elsewhere in the codebase

Reply:
I remove this in the next PR anyway (which introduces the `HeadRelayer`), so going to leave as is for now.